### PR TITLE
Reject null body in feign clients

### PR DESCRIPTION
**Pull request by iamdanfox**
relevant to https://github.com/palantir/http-remoting/issues/758

### CODE REVIEW

File: jaxrs-clients/src/main/java/com/palantir/remoting3/jaxrs/AbstractFeignJaxRsClientBuilder.java [hunk 1]
```diff
@@ -135,6 +136,7 @@ private static Decoder createDecoder(ObjectMapper objectMapper, ObjectMapper cbo
                                 new TextDelegateDecoder(
                                         new CborDelegateDecoder(
                                                 cborObjectMapper,
-                                                new JacksonDecoder(objectMapper))))));
+                                                new RejectNullDecoder(
+                                                        new JacksonDecoder(objectMapper)))))));
```
Comment (dansanduleac) [inline jaxrs-clients/src/main/java/com/palantir/remoting3/jaxrs/AbstractFeignJaxRsClientBuilder.java:140]: moar nesting!!

**Review by dansanduleac — APPROVED** [review]

**Comment by qinfchen** [comment]
i think this violates the wire spec
> Empty or null optional, map, set and list deserialize without error.

**Comment by markelliot** [comment]
what if the main request body is `optional<string>`, for instance?

**Comment by iamdanfox** [comment]
@markelliot the 204 -> optional conversion behaviour all still works, proved by this test: https://github.com/palantir/http-remoting/blob/fc89c8b85182de73ee441faee907052762e1f034/jaxrs-clients/src/test/java/com/palantir/remoting3/jaxrs/feignimpl/Java8OptionalAwareDecoderTest.java#L56

If we also want a HTTP 200 containing the body `null` to deserialize to Optional.empty(), then I need to add that assertion to our master [test-cases.yml](https://github.com/palantir/conjure-verification/blob/develop/test-cases.yml)

**Pull request merged**